Add /mvinv view Command for player inventory inspection across worlds

Woodstop opened the pull request:
This PR introduces the `/mvinv view` and `/mvinv modify` commands, allowing admins to inspect or modify any player's (online or offline) inventory, armor, and off-hand contents in a specific Multiverse world, per suggestion #393. 

# Key Features:

* View Player Inventories: See a player's items across worlds.

* Offline Support: Works for both online and offline players.

* Read-Only GUI: Prevents accidental modification of inventories with `/mvinv view`

* Write-Access GUI: Allows admins to modify inventories with `/mvinv modify`

* Autocompletion: Provides suggestions for player names and worlds.

# Usage:
`/mvinv view <player> <world>`
`/mvinv modify <player> <world>`

# Permission:
`multiverse.inventories.view`
`multiverse.inventories.view.modify`

# Known Issue:
Item Quantity Display Bug: The inventory GUI may sometimes display incorrect quantities for stacked items, particularly armor and weapons, showing '1' instead of the actual stack size. This is a visual bug and does not affect the actual item data.

benwoo1110 reviewed (changes requested):
Thanks for the PR, I haven't compile and test it yet, but I will need to request some general changes to the code structure before I can merge this pr.

1. I encourage splitting longs methods into multiple smaller methods. especially the command method. Maybe even putting the business logic in a separate class e.g. `InvViewProvider`.
2. Does ReadOnlyInventoryHolder and ModifiableInventoryHolder need to be in the listener class? I suggest splitting those into separate classes in a new package maybe called `view`. Then the listener class can be package private instead of public. I want to expose as little implementation details as possible.
3. Make command classes non-public and final, same idea as above.
4. Async loading of inventory data seems to be completely ignored, which doesn't seems to be a good idea, since this feature primarily targets loading of offline player inventory data, and there is no need for data to be immediately served to the user.

benwoo1110 commented:
Also for the known issue on armor and weapons, isn't the allowed stack size for those always only 1? at least without any custom plugin messing with it.

Woodstop commented:
Thanks for the comments. I will work on that and get back to you. 

As for what you said about item stacking, you are correct. Thanks for pointing that out. I had EssentialsX enabled when I was testing, and it like its give command allows those items to stack to 64, which is not the default behavior for vanilla Minecraft. It works as intended when I disable EssentialsX.

Woodstop commented:
I believe I have addressed most if not all of your comments. 

1) I created `InventoryDataProvider` to store the business logic. I realize the view and modify commands are still a bit long especially since I added placeholder items to show where the armor and off-hand slots are in the inventory. I can look into splitting that up into multiple methods if needed.

2) I put those two holders into a `view` package and made main listener class package-private.

3) I made the command classes non-public and final/

4) For online players, I set it to synchronous loading when the online player is currently in the target world. For offline players and for online players who are not in the target world, I set it to asynchronous data loading. The GUI itself opens synchronously.

benwoo1110 commented:
Look a lot better, great work! Also, do you mind I directly commit to this pr? They may be some minor things I want to update.

benwoo1110 reviewed (changes requested):
I have took sometime to analysis the code in detail and test (it's quite a big pr, almost 1k lines). I when ahead to directly commit some changes, mainly just for javadocs api versioning and code readability. So please remember to pull before editing your branch future! As for other concerns that came to my mind when reading the code, I have commented on the related lines directly.
----
I also want to highlight some tips on code readability. This is more of an fyi and learning point, not really anything specific to this PR. Firstly, I suggest trying to avoid a single long method. There is usually ways to spilt them up into small private methods, for example:

 ```java
 // a very long method with many if else
 if (...) {
    // action1
    // ... 
    // ... 
    return ...;
} else {
    // action2
    // ... 
    // ... 
    return ...;
}

// action 1 and 2 can be split into individual private methods
if (...) {
    return action1();
}
return action2();
 ```

Another area is to reduce nesting of loops and if..else conditions. There is usually a way to "invert" the condition by making use of method `return`. This will be even more useful when you have smaller targeted methods. example:

```java
// nested version
void func() {
    if (player.isOnline()) {
        Player player = offlinePlayer.getPlayer();
        if (player != null) {
            // do player action
            if (...) {
                // ...
            }
            return;
        } else {
            // log player null
            return;
        }
    } else {
        // log player offline
    }
}

// improved version
void func() {
    if (!player.isOnline()) {
        // log player offline
        return;
    }
    
    Player player = offlinePlayer.getPlayer();
    if (player == null) {
        // log player null
        return;
    }
    
    // do player action
    if (...) {
        // ...
    }
 }
```
----
And lastly, just a thought, is it possible to also view other non-inventory data such as health, xp, etc.. using the inventory gui, with like lores that you can hover over?

benwoo1110 commented on src/main/java/org/mvplugins/multiverse/inventories/commands/InventoryModifyCommand.java:
```diff
@@ -0,0 +1,146 @@
+                    Bukkit.getScheduler().runTask(inventories, () -> {
+                        // Create inventory with ModifiableInventoryHolder
+                        // Pass all necessary context to the holder for saving on close.
+                        Component title = Component.text("Modify " + targetPlayer.getName() + " @ " + worldName);
```
Oh yea, don't use adventure text api, as we still have to maintain spigot compatibility for now.

Woodstop commented on src/main/java/org/mvplugins/multiverse/inventories/commands/InventoryModifyCommand.java:
```diff
@@ -0,0 +1,146 @@
+            @Description("The world the player's inventory is in")
+            MultiverseWorld world
+    ) {
+        if (player.getUniqueId().equals(targetPlayer.getUniqueId())) {
+            issuer.sendError(ChatColor.RED + "You cannot modify your own inventory using this command. Use your regular inventory.");
+            return;
+        }
```
There's currently a bug where if I were to modify my own live inventory, adding/removing from my GUI inventory will either duplicate or remove the item entirely. In other words, if I take an item out of the GUI inventory and put it into my actual inventory, it will make the item disappear. If I add an item from my actual inventory to the GUI inventory, it will duplicate the item. 

I put this check here to prevent that from happening, but perhaps a better way would be to allow players to modify their own inventories in other worlds, just not their current world. If players can update their stored inventory instead of their live inventory, there shouldn't be any issues. Players can change their inventory normally without this command, so a better use seems to be letting players modify their inventories in other worlds.

Woodstop commented on src/main/java/org/mvplugins/multiverse/inventories/view/ModifiableInventoryHolder.java:
```diff
@@ -0,0 +1,98 @@
+    private final OfflinePlayer targetPlayer;
+    private final String worldName;
+    private final ProfileType profileType;
+    private final MultiverseInventories inventories;
```
You're right, thanks!

Woodstop commented:
I began an attempt at implementing inventory slots to show player data for health, hunger, exp, and last location. I was able to get it working for all but the last location. 

When a player is online, I am using `onlineTarget.getLocation()` to record their location at the time that the GUI is opened. This works fine. However, when trying to access the location data of an offline player or in a player's previous world, it doesn't seem to work right now. I am using `Sharables.LAST_LOCATION` to access the stored location data. Any time I try to access the stored location data, it reads null. I am not sure why, so I left debugging code in the commit in `InventoryDataProvider` for now.

Woodstop commented:
Oh I see that now, thanks! I missed that `active-optional-shares` was an option. I think it's possible that enabling these player-stat item slots could be a config option, mainly so it's clear to people that they need to enable the shareables first. They could also potentially choose what stats they want to see in the slot. 

I'm also fine just leaving it enabled by default even if it shows null, then people can go look at the command documentation to see that they need to enable shareables. There are a lot of ways to handle it

benwoo1110 reviewed (approved):
I will be merging this to our 5.2 branch to be released as the next minor version. Again, thanks for the contribution! (hope to see more in the future too!)

Woodstop commented:
> I will be merging this to our 5.2 branch to be released as the next minor version. Again, thanks for the contribution! (hope to see more in the future too!)

Sounds good! Thanks for the help!
